libsolidity/analysis/ControlFlowRevertPruner.cpp

```diff
@@ -0,0 +1,223 @@
+	{
+		solAssert(*identifier->annotation().requiredLookup == VirtualLookup::Virtual, "");
+		returnFunctionDef = &unresolvedFunctionDefinition.resolveVirtual(*_contract);
+	}
+
+	if (returnFunctionDef && returnFunctionDef->isImplemented())
+		return returnFunctionDef;
+
+	return nullptr;
+}
+
+void ControlFlowRevertPruner::findRevertStates()
+{
+	std::set<CFG::FunctionContractTuple> pending_functions = keys(m_functions);
+
+	while (!pending_functions.empty())
+	{
+		CFG::FunctionContractTuple item = *pending_functions.begin();
+		pending_functions.erase(pending_functions.begin());
+
+		if (m_functions[item] != RevertState::Unknown)
+			continue;
+
+		bool foundExit = false;
+		bool foundUnknown = false;
+
+		FunctionFlow const& functionFlow = m_cfg.functionFlow(*item.function, item.contract);
+
+		solidity::util::BreadthFirstSearch<CFGNode*>{{functionFlow.entry}}.run(
+			[&](CFGNode* _node, auto&& _addChild) {
+				if (_node == functionFlow.exit)
+					foundExit = true;
+
+				for (auto const* functionCall: _node->functionCalls)
+				{
+					auto const* resolvedFunction = resolveCall(*functionCall, item.contract);
+
+					if (resolvedFunction == nullptr)
+						continue;
+
+					switch (m_functions.at({findScopeContract(*resolvedFunction, item.contract), resolvedFunction}))
+					{
+						case RevertState::Unknown:
+							foundUnknown = true;
+							return;
+						case RevertState::AllPathsRevert:
+							return;
+						default:
+							break;
+					}
+				}
+
+				for (CFGNode* exit: _node->exits)
+					_addChild(exit);
+		});
+
+		auto& revertState = m_functions[item];
+		auto previousState = revertState;
+
+		if (foundExit)
+			revertState = RevertState::HasNonRevertingPath;
+		else if (foundUnknown)
+			solAssert(revertState == RevertState::Unknown, "");
+		else
+			revertState = RevertState::AllPathsRevert;
+
+		// Mark all functions depending on this one as modified again
+		if (previousState != revertState)
+			for (auto& nextItem: m_calledBy[item.function])
+				// Ignore different most derived contracts in dependent callees
+				if (
+					item.contract == nullptr ||
+					nextItem.contract == nullptr ||
+					nextItem.contract == item.contract
+				)
+					pending_functions.insert(nextItem);
+	}
+}
+
+void ControlFlowRevertPruner::modifyFunctionFlows()
+{
+	for (auto& item: m_functions)
+	{
+		FunctionFlow const& functionFlow = m_cfg.functionFlow(*item.first.function, item.first.contract);
+		solidity::util::BreadthFirstSearch<CFGNode*>{{functionFlow.entry}}.run(
+			[&](CFGNode* _node, auto&& _addChild) {
+				for (auto const* functionCall: _node->functionCalls)
+				{
+					auto const* resolvedFunction = resolveCall(*functionCall, item.first.contract);
+
+					if (resolvedFunction == nullptr)
+						continue;
+
+					switch (m_functions.at({findScopeContract(*resolvedFunction, item.first.contract), resolvedFunction}))
+					{
+						case RevertState::Unknown:
```

Comment:
Please add a comment either here or somewhere else: If the revert states of the functions do not change anymore, we treat all "unknown" states as "reverting", since they can only be caused by unconditional recursion.

Reply:
We don't actually check the conditions of recursion, so "unconditional" is probably wrong